277 first pass at breaking down API calls and Roles

lewisdaly opened the pull request:
Closes #277

- Lists out all existing FSPIOP API calls, and newly proposed PISP calls
- Update roles doc based for new PISP endpoint: `thirdPartyRequest/transaction`

I'm putting this up as a draft because I'm sure some of these endpoints will change in the next few days with feedback from the CCB and the PISP team.

mjbrichards commented:
First thoughts:

- A general point: in our world, being able to issue a **PUT** in response to a **POST** or a **GET** implies an ability to be open to further interactions. So we shouldn't allow participants to respond unless they're up for it...
- PISP should not be able to respond to a **GET /parties**, since it doesn't own any accounts...
- PISP should not be able to respond to a **POST /thirdPartyRequests/transfers**, since it can't initiate the transfer process.
- I agree that a PISP shouldn't be able to respond to a **POST quotes** request.
- I think our current plan on requesting account balances is that it will be a sub-endpoint of **thirdPartyRequests**: so the second of your alternatives
- We could distinguish between the authorizations forms for DFSP and PISP by allowing only a DFSP to issue a **GET /authorizations**, and only a PISP to issue a **POST /authorizations**

eoln commented on docs/roles.md:
```diff
@@ -0,0 +1,151 @@
+
+## Performance Considerations:
+
+We assume Role Based Access Controls will be implemented at an API Gateway.
```
Is it possible that RBAC could be used not only on APIGateway but also in endpoints implementations to deliver different behavior for different types of Parties (PISP and DFSP)? If yes how the information about PartyType will be passed down?

eoln replied:
Maybe it is possible to implement RBAC on the SDK Scheme Adapter instead of APIGateway? 
https://github.com/mojaloop/sdk-scheme-adapter/ 
It is placed as a protocol adapter between Hub and Parties (DFSP and maybe PISP...)
It is dedicated to performing the operations related to security and header processing.

mjbrichards replied:
The plan is, I think, that the scheme will maintain a central register of which services each participant offers. and this register will be used to implement the RBAC test. We envisage that this will be slowly changing information, and that it will not therefore be necessary for participants to state on each message what service they are requesting. I believe that we use fixed IP addresses, so the RBAC should be able to identify the party directly.

Unfortunately, we don't mandate that participants should make use of the scheme adapter, so it would not be a reliable location for the test.

eoln replied:
thank you @MichaelJBRichards for the explanation. But maybe it would be worth investigating the need for an adapter for PISP?

lewisdaly replied:
@eoln to Michael's point, I think we can safely assume that these access controls lie outside of the scope for what we are doing for the moment. 

As for the adapter for PISP, can you expand a little more on this? Do you mean a sdk-scheme-adapter for PISPs instead of DFSPs?

eoln commented on docs/roles.md:
```diff
@@ -0,0 +1,151 @@
+
+In order to ensure such an implementation won't greatly affect performance, the goal is to have RBAC that need not inspect the _body_ of the API call (or at least minimize the number of API calls that depend on inspecting the body).
+
+This is possible with the PISP Case by using the a new resource for transaction requests: `thirdparty
```
How and where in Hub we plan to implement new `thirdPartyRequests` resource? 
As a core component, or an additional service?
What about the ml-api-adapter? Will it be impacted?

eoln replied:
Any hint how will we implement `consents` and `consentRequests` resources? As one `consentsService`? Will this service be part of the Hub or some adapter?

mjbrichards replied:
The plan is that the Open API definition will be amended to include the new thirdPartyRequests resource. The API adapter will, of course need to be adapted to support it, as is true for any change to the Open API specification.

I think that the plan will also be to add the consents and consentRequest resources to the Open API as well. Our thinking at the moment is that participants will send their requests to the hub and the hub will route them as required, including interacting with a central FIDO server where DFSPs do not implement their own FIDO server.

lewisdaly replied:
> Any hint how will we implement `consents` and `consentRequests` resources? As one `consentsService`? Will this service be part of the Hub or some adapter?

I'm not sure what you mean by part of the Hub _or_ adapter, can you please clarify?

As for the implementation, my gut feeling (which needs further investigation) is that it will be something like a `ConsentsService` as you say, with perhaps a better name, provided that we can separate out its functions into a separate service appropriately.

eoln commented on docs/roles.md:
```diff
@@ -0,0 +1,151 @@
+> New PISP Calls
+- Update Consent Request - `PUT /consentRequests/{ID`
+- Update Consent Request Error - `PUT /consentRequests/{ID}/error`
+- Update Consent - `POST /consents/{ID}`
```
if this is `Inbound`, shouldn't be `PUT` instead of `POST` for `/consents/{ID}`?

mjbrichards replied:
I think that's right: it looks like a repeat of l.47, and there is no substantive response required to this call. As I understand it, it's simply informative; however, we might still require a PUT to confirm to the DFSP that the PISP had registered the permission. Lewis, what do you think?

lewisdaly replied:
Yes I think your're right @eoln.

My original thought here was that the `consent` is created by the DFSP as the result of a successful `consentRequest` (that's the call on L:47, which I have renamed **Create Consent**) and that the PISP has no notion of that created `consent` until it it recieves a `POST`.

If, however, we are happy with `PUT` to act as a Update or Create (which I think we should be, since that is how HTTP works), then this should be:
```
- Create Or Update Consent - `PUT /consents/{ID}`
```

lewisdaly commented:
> PISP should not be able to respond to a **POST /thirdPartyRequests/transfers**, since it can't initiate the transfer process.

Agreed. I think I had a trailing `/{ID}` which implied an update, but I've removed that to simply:
```
- Initiate 3rd party request - `POST /thirdPartyRequest/transaction`
```

_Note: I've gone with `transaction` here, based on the naming conventions of `transactionRequests`, but please let me know if this should be `transfer` instead_

lewisdaly commented:
Hi all, I was getting mightily lost in this document, so I updated it to use a markdown table, and removed the redundancy in API call lists by collapsing them into one list of calls + permissions.

I hope this is more clear now.

lewisdaly commented:
I think this is ready to go for the current state of designs. 

I'm sure there are changes to make down the line, but I don't want to hold this up from getting into the PISP repo where it can be useful for the team/
